In [1]:
# Import Dependencies
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
import warnings
warnings.filterwarnings('ignore')

In [2]:
# Loading the dataset to pandas dataframe
credit_card_data = pd.read_csv('creditcard.csv')

In [3]:
# first five rows
credit_card_data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [4]:
# last five rows
credit_card_data.tail()

Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[5 rows x 31 columns]

In [5]:
# dataset information
credit_card_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [6]:
# checking the number of missing values
credit_card_data.isnull().sum().sum()

0

In [7]:
# distribution of legit transactions & fraudulent transactions
credit_card_data['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

This dataset is highly unbalanced.

0 --> Normal Transactions
1 --> Fraudulent Transactions

In [8]:
# septarating data for analysis
legit = credit_card_data[credit_card_data['Class']==0]
fraud = credit_card_data[credit_card_data['Class']==1]

In [9]:
print(legit.shape)
print(fraud.shape)

(284315, 31)
(492, 31)


In [10]:
# Statistical Measures of the data
legit['Amount'].describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [11]:
fraud['Amount'].describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [12]:
# Compare the values for both transactions
credit_card_data.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

Under-Sampling

Build a sample dataset containing similar distribution of normal transactions and fraudulent transactions

Number of fraudulent transactions --> 492

In [13]:
legit_sample = legit.sample(n=492,random_state=0)

In [14]:
# Concatenating two dataframes
new_dataset = pd.concat([legit_sample, fraud], axis=0)

In [15]:
new_dataset.head()

Time        V1        V2        V3        V4        V5        V6  \
266085  162183.0  2.049094  0.186189 -1.707198  0.530768  0.160589 -1.448570   
172120  120967.0  2.125540 -0.030714 -1.527653  0.121046  0.543172 -0.347988   
15136    26484.0 -4.155859 -5.705748  0.274699 -0.993262 -6.059393  5.210848   
96393    65751.0 -0.566420 -0.579576  0.823503 -1.451240 -0.583587  0.206381   
208225  137025.0  0.060858 -0.261762 -1.699493 -1.202327  3.699527  3.196249   

              V7        V8        V9  ...       V21       V22       V23  \
266085  0.239310 -0.353611  0.634425  ...  0.197782  0.741141 -0.009744   
172120  0.157221 -0.229126  0.477999  ... -0.336497 -0.838932  0.275173   
15136   5.811316  0.367888  1.750710  ...  1.371671  1.195815  4.188762   
96393   1.601392 -0.370446 -1.910354  ... -0.065082 -0.761357  0.641524   
208225  0.437208  0.421541  0.492435  ...  0.008303  0.534602  0.089602   

             V24       V25       V26       V27       V28   Amount  Class  
266085 -0.085057  0.228384 -0.097292 -0.001028 -0.032390     2.99      0  
172120  0.049145 -0.156765  0.205919 -0.072321 -0.059009     1.98      0  
15136  -1.091077  1.033044  0.224493 -0.486741  0.194275  1937.66      0  
96393  -0.568974 -0.053164 -0.690995 -0.228630 -0.157254   320.05      0  
208225  0.667918  0.017798  0.611584 -0.469946 -0.514370    11.50      0  

[5 rows x 31 columns]

In [16]:
new_dataset.tail()

Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[5 rows x 31 columns]

In [17]:
new_dataset['Class'].value_counts()

0    492
1    492
Name: Class, dtype: int64

In [18]:
new_dataset.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      91723.750000  0.075394 -0.081458 -0.015026  0.120566 -0.031588   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.077275  0.039596  0.040481  0.033689  ...  0.016588  0.001340   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.009691  0.006777  0.021083  0.024237 -0.004159 -0.024761  0.013434   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0      104.396870  
1      122.211321  

[2 rows x 30 columns]

The difference is still there. Therefore basic nature of data has not changed.

In [19]:
# Splitting the data into Features and Targets
X = new_dataset.iloc[:,0:-1]
Y = new_dataset.iloc[:, -1]

In [20]:
X.head()

Time        V1        V2        V3        V4        V5        V6  \
266085  162183.0  2.049094  0.186189 -1.707198  0.530768  0.160589 -1.448570   
172120  120967.0  2.125540 -0.030714 -1.527653  0.121046  0.543172 -0.347988   
15136    26484.0 -4.155859 -5.705748  0.274699 -0.993262 -6.059393  5.210848   
96393    65751.0 -0.566420 -0.579576  0.823503 -1.451240 -0.583587  0.206381   
208225  137025.0  0.060858 -0.261762 -1.699493 -1.202327  3.699527  3.196249   

              V7        V8        V9  ...       V20       V21       V22  \
266085  0.239310 -0.353611  0.634425  ... -0.232666  0.197782  0.741141   
172120  0.157221 -0.229126  0.477999  ... -0.125804 -0.336497 -0.838932   
15136   5.811316  0.367888  1.750710  ...  3.944592  1.371671  1.195815   
96393   1.601392 -0.370446 -1.910354  ...  0.977739 -0.065082 -0.761357   
208225  0.437208  0.421541  0.492435  ... -0.014550  0.008303  0.534602   

             V23       V24       V25       V26       V27       V28   Amount  
266085 -0.009744 -0.085057  0.228384 -0.097292 -0.001028 -0.032390     2.99  
172120  0.275173  0.049145 -0.156765  0.205919 -0.072321 -0.059009     1.98  
15136   4.188762 -1.091077  1.033044  0.224493 -0.486741  0.194275  1937.66  
96393   0.641524 -0.568974 -0.053164 -0.690995 -0.228630 -0.157254   320.05  
208225  0.089602  0.667918  0.017798  0.611584 -0.469946 -0.514370    11.50  

[5 rows x 30 columns]

In [21]:
Y.head()

266085    0
172120    0
15136     0
96393     0
208225    0
Name: Class, dtype: int64

In [22]:
# Splitting the data into training data and testing data
x_train, x_test , y_train, y_test = train_test_split(X,Y,test_size=0.2,stratify=Y,random_state=2)

In [23]:
# Model Creation

model = LogisticRegression()

In [24]:
# Training the model
model.fit(x_train, y_train)

LogisticRegression()

In [25]:
# Predicting values based on testing data

y_pred = model.predict(x_test)

In [26]:
y_pred

array([1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0,
       0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1,
       1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 0,
       1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0,
       0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0,
       1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1,
       0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1,
       0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0],
      dtype=int64)

In [27]:
# Model Evaluation

accuracy_score(y_test, y_pred)

0.934010152284264

Accuracy Score is 93.40%.

In [28]:
confusion_matrix(y_test, y_pred)

array([[95,  4],
       [ 9, 89]], dtype=int64)

In [29]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.91      0.96      0.94        99
           1       0.96      0.91      0.93        98

    accuracy                           0.93       197
   macro avg       0.94      0.93      0.93       197
weighted avg       0.94      0.93      0.93       197



Model Precision --> 96%

Model Recall --> 91%

Accuracy, Precision and Recall, all are above 90% which demonstrates that this model is highly accurate in identifying the legit and fraudulent credit card transactions.